File: sdk/engine/src/conversation/__tests__/manager.test.ts
import { describe, it, expect } from 'vitest';
import { ConversationManager } from '../manager.js';
import { MemoryStorage } from '../storage.js';

function createMgr() {
  return new ConversationManager(new MemoryStorage());
}

describe('ConversationManager', () => {
  it('creates and retrieves a conversation', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', 'you are a bot');

    expect(conv.id).toBeTruthy();
    expect(conv.title).toBe('test');
    expect(conv.messages).toHaveLength(1);
    expect(conv.messages[0].role).toBe('system');

    const got = mgr.get(conv.id);
    expect(got).not.toBeNull();
    expect(got!.id).toBe(conv.id);
  });

  it('creates without system prompt', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    expect(conv.messages).toHaveLength(0);
  });

  it('lists conversations', () => {
    const mgr = createMgr();
    mgr.create('a', '');
    mgr.create('b', '');
    expect(mgr.list()).toHaveLength(2);
  });

  it('deletes a conversation', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    mgr.delete(conv.id);
    expect(mgr.get(conv.id)).toBeNull();
  });

  it('updates messages', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    mgr.updateMessages(conv.id, [
      { role: 'user', content: 'hi' },
      { role: 'assistant', content: 'hello' },
    ]);
    const got = mgr.get(conv.id)!;
    expect(got.messages).toHaveLength(2);
  });

  it('edits a message and truncates', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', 'sys');
    mgr.updateMessages(conv.id, [
      { role: 'system', content: 'sys' },
      { role: 'user', content: 'old' },
      { role: 'assistant', content: 'reply' },
    ]);

    mgr.editMessage(conv.id, 1, 'new');
    const got = mgr.get(conv.id)!;
    expect(got.messages).toHaveLength(2);
    expect(got.messages[1].content).toBe('new');
  });

  it('deletes a single message', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    mgr.updateMessages(conv.id, [
      { role: 'user', content: 'a' },
      { role: 'assistant', content: 'b' },
      { role: 'user', content: 'c' },
    ]);

    mgr.deleteMessage(conv.id, 1);
    const got = mgr.get(conv.id)!;
    expect(got.messages).toHaveLength(2);
    expect(got.messages[1].content).toBe('c');
  });

  it('deletes messages from index', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    mgr.updateMessages(conv.id, [
      { role: 'user', content: 'a' },
      { role: 'assistant', content: 'b' },
      { role: 'user', content: 'c' },
    ]);

    mgr.deleteMessagesFrom(conv.id, 1);
    const got = mgr.get(conv.id)!;
    expect(got.messages).toHaveLength(1);
  });

  it('updates tool config', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    mgr.updateToolConfig(conv.id, ['github'], ['delete_repo']);

    const got = mgr.get(conv.id)!;
    expect(got.enabledSources).toEqual(['github']);
    expect(got.disabledTools).toEqual(['delete_repo']);
  });

  it('inherits tool config from latest conversation', () => {
    const mgr = createMgr();
    const conv1 = mgr.create('first', '');
    mgr.updateToolConfig(conv1.id, ['github'], ['delete_repo']);

    const conv2 = mgr.create('second', '');
    expect(conv2.enabledSources).toEqual(['github']);
  });

  it('throws on edit with invalid index', () => {
    const mgr = createMgr();
    const conv = mgr.create('test', '');
    expect(() => mgr.editMessage(conv.id, 5, 'x')).toThrow();
  });

  it('throws on edit with missing conversation', () => {
    const mgr = createMgr();
    expect(() => mgr.editMessage('nonexistent', 0, 'x')).toThrow();
  });

  it('updates title', () => {
    const mgr = createMgr();
    const conv = mgr.create('old', '');
    mgr.updateTitle(conv.id, 'new');
    expect(mgr.get(conv.id)!.title).toBe('new');
  });
});

File: core/conversation/manager_test.go
package conversation

import (
	"testing"

	"github.com/ZacharyZcR/NLUI/core/llm"
)

func TestCreateAndGet(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "you are a bot")

	if conv.ID == "" {
		t.Fatal("ID should not be empty")
	}
	if conv.Title != "test" {
		t.Errorf("title = %q, want test", conv.Title)
	}
	if len(conv.Messages) != 1 || conv.Messages[0].Role != "system" {
		t.Error("should have system message")
	}

	got := m.Get(conv.ID)
	if got == nil {
		t.Fatal("Get returned nil")
	}
	if got.ID != conv.ID {
		t.Error("ID mismatch")
	}
}

func TestCreateWithoutSystemPrompt(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	if len(conv.Messages) != 0 {
		t.Errorf("expected 0 messages, got %d", len(conv.Messages))
	}
}

func TestList(t *testing.T) {
	m := NewManager("")
	m.Create("a", "")
	m.Create("b", "")
	list := m.List()
	if len(list) != 2 {
		t.Fatalf("expected 2, got %d", len(list))
	}
}

func TestDelete(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	m.Delete(conv.ID)
	if m.Get(conv.ID) != nil {
		t.Fatal("should be deleted")
	}
}

func TestUpdateMessages(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	msgs := []llm.Message{{Role: "user", Content: "hi"}, {Role: "assistant", Content: "hello"}}
	m.UpdateMessages(conv.ID, msgs)

	got := m.Get(conv.ID)
	if len(got.Messages) != 2 {
		t.Fatalf("expected 2 messages, got %d", len(got.Messages))
	}
}

func TestEditMessage(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "sys")
	m.UpdateMessages(conv.ID, []llm.Message{
		{Role: "system", Content: "sys"},
		{Role: "user", Content: "old"},
		{Role: "assistant", Content: "reply"},
	})

	err := m.EditMessage(conv.ID, 1, "new")
	if err != nil {
		t.Fatal(err)
	}

	got := m.Get(conv.ID)
	if len(got.Messages) != 2 {
		t.Fatalf("expected 2 messages (truncated after edit), got %d", len(got.Messages))
	}
	if got.Messages[1].Content != "new" {
		t.Errorf("content = %q, want new", got.Messages[1].Content)
	}
}

func TestDeleteMessage(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	m.UpdateMessages(conv.ID, []llm.Message{
		{Role: "user", Content: "a"},
		{Role: "assistant", Content: "b"},
		{Role: "user", Content: "c"},
	})

	err := m.DeleteMessage(conv.ID, 1)
	if err != nil {
		t.Fatal(err)
	}

	got := m.Get(conv.ID)
	if len(got.Messages) != 2 {
		t.Fatalf("expected 2, got %d", len(got.Messages))
	}
	if got.Messages[1].Content != "c" {
		t.Errorf("got %q, want c", got.Messages[1].Content)
	}
}

func TestDeleteMessagesFrom(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	m.UpdateMessages(conv.ID, []llm.Message{
		{Role: "user", Content: "a"},
		{Role: "assistant", Content: "b"},
		{Role: "user", Content: "c"},
	})

	err := m.DeleteMessagesFrom(conv.ID, 1)
	if err != nil {
		t.Fatal(err)
	}

	got := m.Get(conv.ID)
	if len(got.Messages) != 1 {
		t.Fatalf("expected 1, got %d", len(got.Messages))
	}
}

func TestUpdateToolConfig(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	err := m.UpdateToolConfig(conv.ID, []string{"github"}, []string{"delete_repo"})
	if err != nil {
		t.Fatal(err)
	}

	got := m.Get(conv.ID)
	if len(got.EnabledSources) != 1 || got.EnabledSources[0] != "github" {
		t.Errorf("enabled_sources = %v", got.EnabledSources)
	}
	if len(got.DisabledTools) != 1 || got.DisabledTools[0] != "delete_repo" {
		t.Errorf("disabled_tools = %v", got.DisabledTools)
	}
}

func TestToolConfigInherited(t *testing.T) {
	m := NewManager("")
	conv1 := m.Create("first", "")
	m.UpdateToolConfig(conv1.ID, []string{"github"}, []string{"delete_repo"})

	conv2 := m.Create("second", "")
	if len(conv2.EnabledSources) != 1 || conv2.EnabledSources[0] != "github" {
		t.Errorf("should inherit enabled_sources, got %v", conv2.EnabledSources)
	}
}

func TestEditMessageOutOfBounds(t *testing.T) {
	m := NewManager("")
	conv := m.Create("test", "")
	err := m.EditMessage(conv.ID, 5, "x")
	if err == nil {
		t.Fatal("expected error for out of bounds")
	}
}

func TestEditMessageNotFound(t *testing.T) {
	m := NewManager("")
	err := m.EditMessage("nonexistent", 0, "x")
	if err == nil {
		t.Fatal("expected error for not found")
	}
}
